store.go

```diff
@@ -2938,6 +2984,19 @@ func (s *store) UnmountImage(id string, force bool) (bool, error) {
 	if err != nil {
 		return false, err
 	}
+	// check if the layer with the canonical mapping is in the mapped top layers
+	for _, layerID := range img.MappedTopLayers {
+		l, err := s.Layer(layerID)
```

Comment:
This works, but it takes O(`len(MappedTopLayers)*len(roImageStores)`) locks, _and_ it can later fail due to a race with a “layer unknown”.

Consider (adding a comment to note that `MountImage` only creates mounts in the primary store and) manually locking the primary store, and looking up layers only in that store. That would ~require inlining `s.Unmount`.